File: transpiler/javatests/com/google/j2cl/transpiler/integration/morebridgemethods/OverwrittenTypeVariablesMain.java
package com.google.j2cl.transpiler.integration.morebridgemethods;

enum A {
  FIRST,
  SECOND;
}

interface MyFunction<F, T> {
  T apply(F input);
}

interface HashFunction<T> extends MyFunction<T, String> {}

class HashFunctions {
  public static final <T extends Enum<T>> HashFunction<T> enumHashFunction() {
    return new HashFunction<T>() {
      @Override
      public String apply(T input) {
        return "" + input.ordinal();
      }
    };
  }
}

public class OverwrittenTypeVariablesMain {
  public static void test() {
    // Might crash.
    HashFunctions.enumHashFunction();
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/morebridgemethods/OverwrittenTypeVariablesMain.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package morebridgemethods

internal enum class A {
  FIRST,
  SECOND
}

internal interface MyFunction<F, T> {
  fun apply(input: F): T
}

internal interface HashFunction<T> : MyFunction<T, String>

internal object HashFunctions {
  fun <T : Enum<T>> enumHashFunction(): HashFunction<T> {
    return object : HashFunction<T> {
      override fun apply(input: T): String {
        return "" + input.ordinal
      }
    }
  }
}

object OverwrittenTypeVariablesMain {
  fun test() {
    // Might crash.
    HashFunctions.enumHashFunction()
  }
}
